R/read-write.R

```diff
@@ -0,0 +1,465 @@
+    if (!all(col_class_tibble$class %in% xlsx_classes)) {
+      abort_class_not_supported(setdiff(col_class_tibble$class, xlsx_classes))
+    }
+  }
+
+  # FIXME: might need to revisit for compound keys
+  pk_tibble <- dm_get_all_pks_impl(dm)
+
+  # not using dm_get_all_fks_impl(), because it will break with introduction of compound keys
+  # FIXME: might need to revisit for compound keys
+  fk_tibble <- dm_get_def(dm) %>%
+    select("child_table" = table, fks) %>%
+    unnest(fks) %>%
+    rename(parent_table = table, fk_col = column) %>%
+    # key number is needed in case of compound keys
+    # FIXME: referenced column of PK then needs to be given too
+    mutate(key_nr = row_number()) %>%
+    unnest(fk_col)
+
+  list(
+    info_tibble = info_tibble,
+    col_class_tibble = col_class_tibble,
+    pk_tibble = pk_tibble,
+    fk_tibble = fk_tibble
+  )
+}
+
+dm_write_csv_impl <- function(dm, csv_directory, zip) {
+  csv_tables <- prepare_tbls_for_def_and_class(dm, csv = TRUE)
+
+  if (!dir.exists(csv_directory)) {
+    dir.create(csv_directory)
+  }
+
+  csv_info_filename <- file.path(csv_directory, "___info_file_dm.csv")
+  csv_coltypes_filename <- file.path(csv_directory, "___coltypes_file_dm.csv")
+  csv_pk_filename <- file.path(csv_directory, "___pk_file_dm.csv")
+  csv_fk_filename <- file.path(csv_directory, "___fk_file_dm.csv")
+  csv_table_filenames <- file.path(csv_directory, paste0(src_tbls(dm), ".csv"))
+
+  readr::write_csv(csv_tables$info_tibble, csv_info_filename)
+  readr::write_csv(csv_tables$col_class_tibble, csv_coltypes_filename)
+  readr::write_csv(csv_tables$pk_tibble, csv_pk_filename)
+  readr::write_csv(csv_tables$fk_tibble, csv_fk_filename)
+
+  dm_tables <- dm_get_tables_impl(dm) %>%
+    convert_all_times_to_utc(csv_tables$col_class_tibble)
+  walk2(dm_tables, csv_table_filenames, readr::write_csv)
+  if (!zip) {
+    message(
+      glue::glue("Written `dm` as collection of csv-files to directory {tick(csv_directory)}.")
+    )
+  }
+  invisible(dm)
+}
+
+dm_col_class <- function(dm) {
+  tables <- dm_get_tables_impl(dm)
+  map(dm_get_tables_impl(dm), colnames) %>%
+    imap_dfr(function(col, table) {tibble(table = table, column = col)}) %>%
+    mutate(class = map2_chr(
+      table,
+      column,
+      function(table, column) {
+        tables[[table]] %>%
+          pull(column) %>%
+          class() %>%
+          pluck(1)}
+      )
+    )
+}
+
+readr_translate <- function(r_class) {
+  map_chr(r_class, ~ filter(class_translation_r_readr, r_class == .x) %>% pull(readr_code))
+}
+
+class_translation_r_readr <- tribble(
+  ~r_class, ~readr_code,
+  "logical", "l",
+  "integer", "i",
+  "numeric", "d",
+  "character", "c",
+  "Date", "D",
+  "POSIXlt", "T",
+  "POSIXct", "T"
+)
+
+xlsx_classes <- c(
+  "logical",
+  "integer",
+  "numeric",
+  "character",
+  "Date",
+  "POSIXlt",
+  "POSIXct"
+)
+
+convert_all_times_to_utc <- function(table_list, col_class_table) {
```

Comment:
Is this within the scope of the function? Will it affect the roundtrip, or only the snapshot tests, if we omit UTC conversion?

Reply:
Several problems:
1. unfortunately it's not possible to read timezones via `readr::read_csv()`. Therefore, it seems much safer to convert to UTC and to inform when writing the csv files.
1. for xlsx: actually `writexl::write_xlsx()` does the conversion to UTC quietly itself anyway. In this case we would not need to perform the conversion, but just inform the user.

I think it's not harmful to leave it as is, but I am open to suggestions.

Reply:
> 1. unfortunately it's not possible to read timezones via readr::read_csv(). Therefore, it seems much safer to convert to UTC and to inform when writing the csv file

I was too quick to claim that:
https://readr.tidyverse.org/articles/locales.html
It is actually possible to steer the timezone with the argument `locale` in `readr::read_csv()`. Not sure if making use of this possibility improves the transparency of our functions (mainly if it doesn't work with xlsx)